### ISSUE TITLE
What do I give to screenDPI to get consistently-sized text?

### ISSUE THREAD

**Issue by iffy** [player-03/advanced-layout#6]
(If this is the wrong place to post this, I'm happy to go where I'm supposed to)

I have the following:

`Source/Main.hx`:

```haxe
package;

import openfl.display.Sprite;
import openfl.Assets;
import openfl.text.TextField;
import openfl.text.TextFormat;
import openfl.system.System;
import openfl.system.Capabilities;
import layout.Layout;

using layout.LayoutCreator;


class Main extends Sprite {
    // private var _conn:Connection = null;
    
    public function new() {
        super();

        // Scale
        trace('stage.window.scale ${stage.window.scale}');
        trace('Capabilities.screenDPI ${Capabilities.screenDPI}');
        Layout.stageLayout.scale.screenDPI(Std.int(Capabilities.screenDPI));

        // Load fonts
        var bold14Font = new TextFormat("assets/fonts/Lato-Bold", 14, 0xffffff);

        // Components
        var navbarBackground = new Sprite();
        addChild(navbarBackground);

        navbarBackground.graphics.beginFill(0x2c3e50);
        navbarBackground.graphics.drawRect(0, 0, 10, 10);

        var text1 = new TextField();
        text1.defaultTextFormat = bold14Font;
        text1.text = 'Hello';
        addChild(text1);

        // Scale components
        navbarBackground.simpleScale();
        navbarBackground.fillHeight();
        navbarBackground.alignTopLeft();
        navbarBackground.exactWidth(100);
    }
}
```

`project.xml`:

```xml
<?xml version="1.0" encoding="utf-8"?>
<project>
    <meta title="Test" package="com.example.test" version="0.1.0" company="Test" />
    <app main="Main" path="Export" file="Test" />
    <source path="Source" />
    <haxelib name="openfl" />
    <haxelib name="advanced-layout" />
    <assets path="Assets" rename="assets" />

    <window allow-high-dpi="true" />
</project>
```

Which renders the textfield and blue bar too small:

<img width="132" alt="screen shot 2018-08-14 at 4 15 25 pm" src="https://user-images.githubusercontent.com/161727/44121445-47472b02-9fdd-11e8-876c-f0ed53ea6ea1.png">

---

If I remove `<window allow-high-dpi="true" />` from `project.xml` it looks like this:

<img width="191" alt="screen shot 2018-08-14 at 4 14 54 pm" src="https://user-images.githubusercontent.com/161727/44121432-34f72100-9fdd-11e8-922b-ac39d21cb3d0.png">

This is the size "in inches" that I want for the blue bar and text, but with blurry text.

---

To make the text sharp, I can add `<window allow-high-dpi="true" />` back in and then add this code.

```haxe
        text1.scaleX = 2;
        text1.scaleY = 2;
```

<img width="151" alt="screen shot 2018-08-14 at 4 13 11 pm" src="https://user-images.githubusercontent.com/161727/44121371-fb3d3ad0-9fdc-11e8-853d-3c3a660623b6.png">

but that doesn't scale the blue bar the same as the text.

---

Changing `Layout.stageLayout.scale.screenDPI(...)` doesn't do anything as far as I can tell.

Is there a way to make the blue bar an "inch" wide on all screens and the text smooth using Advanced Layout?

**Comment by player-03** [comment]
Under the hood, it uses [`Capabilities.screenDPI`](https://github.com/openfl/openfl/blob/develop/src/openfl/system/Capabilities.hx#L302), and I'm guessing that's the problem.

According to [word of Joshua](http://community.openfl.org/t/why-is-capabilities-screendpi-different-from-stage-window-display-dpi-what-should-i-use/9481/2), `Capabilities.screenDPI` "roughly estimates the density you might want to apply to your UI," because "users with higher resolution screens may expect things to look smaller."

He also mentions that [`Display.dpi`](https://github.com/openfl/lime/blob/develop/src/lime/system/Display.hx#L30) calculates it as accurately as possible, with no adjustments made after the fact.

**Comment by player-03** [comment]
Just added an option to choose between the two: c860b191232880eeeae207e9329adcaa60b8877b

**Comment by iffy** [comment]
@player-03 wow, thank you!  What version of lime do I need?  I'm trying with openfl 8.4.1 and lime 7.0.0 and getting this error:

```
/Users/matt/lib/advanced-layout/com/player03/layout/Scale.hx:265: characters 46-57 : Class<lime.system.Display> has no field dpi
```

Here's everything:

```
$ haxelib list
actuate: [1.8.7]
advanced-layout: 0.7.0 [dev:/Users/matt/lib/advanced-layout]
box2d: [1.2.3]
haxeui-core: [0.0.4]
haxeui-openfl: [git]
haxeui: [1.8.21]
hscript: [2.2.0]
hxcpp: 3.4.188 [4.0.4]
layout: [1.2.1]
lime-samples: [6.2.0]
lime: 2.9.1 6.4.0 [7.0.0]
openfl-samples: [6.0.0]
openfl: 3.6.1 8.2.2 8.3.0 8.4.0 [8.4.1]
sqlite: [1.0.9]
```

**Comment by player-03** [comment]
Oops! Somehow I didn't notice that the variable wasn't static. This should be fixed as of 6f5048d427109a8f0a7a846c1b6ee6c73e5409ab.

**Comment by iffy** [comment]
I think I'm not calling `screenDPI(...)` correctly.  Neither changing `baseDPI` nor `smart` seems to have any effect.  Am I calling `screenDPI(...)` on the right instance in my code?

```
import layout.Layout;
...
Layout.stageLayout.scale.screenDPI(Std.int(Capabilities.screenDPI));
```

For reference, here's what I'm seeing:

`screenDPI(144, false)`
<img width="91" alt="screendpi144_false" src="https://user-images.githubusercontent.com/161727/44475657-ddffd600-a5f2-11e8-86f5-2843b8584f56.png">

`screenDPI(144, true)`
<img width="101" alt="screendpi144_true" src="https://user-images.githubusercontent.com/161727/44475714-fff95880-a5f2-11e8-97b0-7b08765b5b70.png">

`screenDPI(72, false)`
<img width="120" alt="screendpi72_false" src="https://user-images.githubusercontent.com/161727/44475722-05ef3980-a5f3-11e8-8d1f-1cdba92b2d85.png">

`screenDPI(72, true)`
<img width="117" alt="screendpi72_true" src="https://user-images.githubusercontent.com/161727/44475730-0be51a80-a5f3-11e8-861d-c6b8078c9a62.png">

**Comment by player-03** [comment]
`Layout.stageLayout.scale.screenDPI()` is correct, but you definitely don't want to pass `Capabilities.screenDPI` as the first argument. Since it divides `Capabilities.screenDPI` by whatever `baseDPI` is, you'll end up with a scale of 1 no matter what. (Also, I just noticed that your original code does the same thing. Should have looked more closely!)

Any of your other four examples ought to work, or at least ought to have some effect. I mean, they worked for me when I added the following code to the end of DisplayingABitmap's `Main` class:

```haxe
//Uncomment for a smaller bitmap:
//Layout.stageLayout.scale.screenDPI(144);

//Uncomment for a larger bitmap:
//Layout.stageLayout.scale.screenDPI(72);

bitmap.simpleScale();
bitmap.centerX();
bitmap.centerY();
```
